phastSim/phastSim.py

```diff
@@ -641,23 +809,196 @@ def sampleMutation(self, allele, rates, rand):
+            node.insertionRate = self.insertionRate[-1]
+            node.deletionRate = self.deletionRate[-1]
+            node.deletionLength = self.deletionLength[-1]
+            node.insertionLength = self.insertionLength[-1]
+
+            node.rate = -1 * node.rates[node.allele][node.allele]
+            
+
+
+            return node
+
+        else:
+            # cut the list in half and build a tree for each half
+            middle = pos_left + int((pos_right - pos_left) / 2)
+            half_length = int((pos_right - pos_left)/2)
+
+            child0 = self.buildTree(genomeNode(level=node.level), insert_list[:half_length+1], pos_left, middle)
+            child1 = self.buildTree(genomeNode(level=node.level), insert_list[half_length+1:], middle+1, pos_right)
+            node.belowNodes = [child0, child1]
+            node.rate = child0.rate + child1.rate
+            node.genomePos = [(self.n_insertions, pos_left), (self.n_insertions, pos_right)]
+            return node
+
+
+    def sampleInsertion(self, rate, node, rand, level):
+
+        # sample a geometric random variable
+        insertion_length = 1 + int(np.log(rand)/np.log(rate))
+        
+        # TODO generate a nucleotide sequence using the correct model
+        insertion = np.random.choice(self.allelesList, insertion_length, self.insertionFrequencies)
+
+        print("*************", insertion)
+        # turn the nucleotide sequence into a genomeTree
+        genomeSeq = "".join(insertion)
+
+        newNode = genomeNode(level=level)
+
+        # generate the genome tree of this insert from (node, genomeSeq)
+        subTree = self.buildTree(newNode, genomeSeq, 0, len(genomeSeq) - 1)
+
+        self.normalizeRates(subTree, self.indels, self.codon, len(genomeSeq), self.scale)
+
+        # build a bigger tree of the form t = (node, subTree). 
+        newRoot = genomeNode(level=level)
+        newRoot.belowNodes = [node, subTree]
+        newRoot.rate = node.rate + subTree.rate
+
+
+        mutData = mutation(mType=mType.INS, genomePos=node.refNode.genomePos[0], insert=genomeSeq, index=self.n_insertions)
+
+        return mutData, newRoot
 
     def findPos(self, rand, parentGenomeNode, level):
         # find position to mutate along the genome, and update temporary genome tree structure as you go
         if parentGenomeNode.isTerminal:
             # reached a terminal node, now sample the mutation event at the position and update all rates
             node = parentGenomeNode.refNode
             a = parentGenomeNode.allele
+            
+            # we'll choose an indel if rand is less than a threshold determined by the insertion and deletion rates for this site
+            if self.indels:
+
+                # deletions
+                if rand < node.deletionRate:
+
+                    attempted_deletion_length = self.sampleDeletion(node.deletionLength, rand/node.deletionRate)
```

Comment:
Why is sampleDeletion() called on node instead of parentGenomeNode ? 

Also, on a completely separate note, I haven't checked what happens when sampleDeletion() is called on a node at the end of the genome, and the deletion length exceeds what remains of the genome. Hopefully nothing weird happens, but it's good to check.

Reply:
For the first question, it's because parentGenomeNode doesn't have any properties except isTerminal and level. But maybe it should have those set... but I am not sure how that would work?

I've answered the second question in another comment below - I will try to check that it works how I have intended!